polling_stations/apps/data_finder/helpers.py

```diff
@@ -7,99 +7,222 @@
+    def geocode(self):
+        addresses = Address.objects.filter(postcode=self.postcode)
+        if not addresses:
+            raise ObjectDoesNotExist('No addresses found for postcode %s' % (self.postcode))
+
+        codes = self.get_codes(self.get_uprns(addresses))
+        centre = centre_from_points_qs(addresses)
+        return {
+            'source': 'addressbase',
+            'wgs84_lon': centre.x,
+            'wgs84_lat': centre.y,
+            'council_gss': codes['council_gss'],
+            'gss_codes': codes['gss_codes'],
+        }
 
-        try:
-            # attempt to parse error from json
-            res_json = res.json()
-            if 'error' in res_json:
-                raise PostcodeError("Mapit error {}: {}".format(res_json['code'], res_json['error']))
-            else:
-                raise PostcodeError("Mapit error {}: unknown".format(res.status_code))
-        except ValueError:
-            # if we fail to parse json, raise a less specific exception
-            raise PostcodeError("Mapit error {}: unknown".format(res.status_code))
+    def geocode_point_only(self):
+        addresses = Address.objects.filter(postcode=self.postcode)
+        if not addresses:
+            raise ObjectDoesNotExist('No addresses found for postcode %s' % (self.postcode))
 
-    res_json = res.json()
+        centre = centre_from_points_qs(addresses)
+        return {
+            'source': 'addressbase',
+            'wgs84_lon': centre.x,
+            'wgs84_lat': centre.y,
+        }
+
+
+def geocode_point_only(postcode, sleep=True, mapit=mapit_wrapper_factory):
+    addressbase = AddressBaseWrapper(postcode)
+    mapit = mapit(postcode)
+
+    try:
+        # first try addressbase
+        result = addressbase.geocode_point_only()
+    except ObjectDoesNotExist:
+        # we couldn't find this postcode in AddressBase: fall back to mapit
+
+        # optional sleep to avoid hammering mapit
+        if sleep:
+            time.sleep(1.3)
+
+        result = mapit.geocode()
+    except:
+        # something else went wrong: lets give mapit a go anyway
+
+        # optional sleep to avoid hammering mapit
+        if sleep:
+            time.sleep(1.3)
 
-    gss_codes = []
-    council_gss = None
-    for area in res_json['areas']:
-        if 'gss' in res_json['areas'][area]['codes']:
-            gss = res_json['areas'][area]['codes']['gss']
-            gss_codes.append(gss)
-            if res_json['areas'][area]['type'] in COUNCIL_TYPES:
-                council_gss = gss
-
-    if 'wgs84_lon' not in res_json or 'wgs84_lat' not in res_json:
-        raise PostcodeError("No location information")
-
-    return {
-        'wgs84_lon': res_json['wgs84_lon'],
-        'wgs84_lat': res_json['wgs84_lat'],
-        'gss_codes': gss_codes,
-        'council_gss': council_gss,
-    }
+        result = mapit.geocode()
+
+    return result
+
+
+def geocode(postcode, mapit=mapit_wrapper_factory):
```

Comment:
(Comment on this line, but really about most of ad33cd9208186b436d64972031ab72a31a39755c)

Few things: 

1. I'm not sure what `mapit_wrapper_factory` is doing, or why `AddressBaseWrapper` doesn't have one. I think `mapit_wrapper_factory` can be removed and `MapitWrapper` called from `geocode` directly.

2. I think the approach here should be to have a list of classes that can do the geocoding, in the order we want. The first class that can do the geocoding is returned. This is more like the pattern I've written here: https://github.com/DemocracyClub/EveryElection/blob/master/every_election/apps/elections/query_helpers.py. 

3. If you think 2. is a good idea, then I think there should be a base class that sets `self.postcode` and formats it and provides a runner something like:

```python
class BaseGeoCoder:
    def __init__(self, postcode):
        self.postcode = self.format_postcode(postcode)
        
    def run(self, point_only=False):
        if point_only:
            return self.geocode_point_only()
        else:
            return self.geocode()
```

Then `geocode` becomes something like this:

```python
def geocode(postcode, point_only=False):
    for geocoder in (AddressBaseGeocoder, MapitGeocoder):
        try:
            return geocoder.run(postcode, point_only)
        except GeocodeError:
            continue

def geocode_point_only(postcode):
    return geocode(postcode, point_only=True)
```

(assume better Exception handling here, but you get the point)

I think this is simpler, and might allow for adding more geocoders in future (for example, we can fall back from AddressBase to the S3 cache to MapIt when we're doing a point_only call.